package/resource_monitor/resource_monitor/src/query_process_file_descriptor.c

```diff
@@ -0,0 +1,456 @@
+                                                .type = FT_STR,
+                                                .dst.str = file_str,
+                                                .buflen = sizeof(file_str),
+                                                .desc = "file descriptor",
+                                                .next = FD_STATE_DONE,
+                                              },
+                                            [FD_STATE_DONE] = (line_spec_t){
+                                              .desc = "done",
+                                            }};
+
+  if (!parse_tab_line(line, FD_STATE_ID_STR, FD_STATE_DONE, line_specs, "\t")) {
+#ifdef DEBUG_QUERY_FD_TAB
+    PK_LOG_ANNO(LOG_ERR, "file descriptor query: parse_tab_line failed ");
+#endif
+    return false;
+  }
+  return process_fd_entry(state, file_str, id_str);
+}
+
+static void destroy_table_entry(table_t *table, void **entry)
+{
+  (void)table;
+  free(*entry);
+  *entry = NULL;
+}
+
+static int compare_counts(const void *ptc1, const void *ptc2)
+{
+  const id_to_count_t *id_to_count1 = ptc1;
+  const id_to_count_t *id_to_count2 = ptc2;
+  return -((int)id_to_count1->count - (int)id_to_count2->count);
+}
+
+static void run_resource_query(void *context)
+{
+  resq_state_t *state = context;
+  if (state->pid_table != NULL) {
+    table_destroy(&state->pid_table);
+  }
+  if (state->file_table != NULL) {
+    table_destroy(&state->file_table);
+  }
+  char leftover_buffer[4096] = {0};
+  leftover_t NESTED_AXX(leftover) = {.buf = leftover_buffer};
+  state->send_state = SEND_FD_COUNTS_0;
+  state->pid_table = table_create(MAX_PROCESS_COUNT, destroy_table_entry);
+  state->file_table = table_create(MAX_FILE_COUNT, destroy_table_entry);
+  state->total_fd_count = 0;
+  if (state->pid_table == NULL) {
+    PK_LOG_ANNO(LOG_ERR, "file descriptor query: pid table creation failed");
+    return;
+  }
+  if (state->file_table == NULL) {
+    PK_LOG_ANNO(LOG_ERR, "file descriptor query: file table creation failed");
+    return;
+  }
+  const char *argv[] = {"sudo", "/usr/bin/lsof", NULL};
+  char buf[LSOF_CMD_BUF_LENGTH] = {0};
+  int rc = run_with_stdin_file(NULL, "sudo", argv, buf, sizeof(buf));
+  if (rc != 0) {
+#ifdef DEBUG_QUERY_FD_TAB
+    PK_LOG_ANNO(LOG_ERR,
+                "file descriptor query: error running 'lsof' command: %s",
+                strerror(errno));
+#endif
+    return;
+  }
+  char *line_ctx = NULL;
+  line_fn_t line_func = NESTED_FN(bool, (const char *line), {
+    if (!parse_fd_line(state, line)) {
+#ifdef DEBUG_QUERY_FD_TAB
+      PK_LOG_ANNO(LOG_ERR, "file descriptor query: parse_fd_line failed ");
+#endif
+      return false;
+    }
+    return true;
+  });
+  ssize_t consumed = foreach_line(buf, &leftover, line_func);
+  if (consumed < 0) {
+    PK_LOG_ANNO(LOG_WARNING, "there was an error parsing the 'ss' output");
+    return;
+  }
+
+
+  for (char *line = strtok_r(buf, "\n", &line_ctx); line != NULL;
```

Comment:
It looks like this for loop is duplicating the work of the `foreach_line` above, should it be removed?